Updated tutorial (#484)
spatialos/UnrealGDK: SpatialGDK/Documentation/content/get-started/tutorial.md

3. Back in your two Unreal game clients, run around and shoot.
4. Using the Inspector to track the location of your two players, notice that if you position them in the area of authority then their shots damage each other, but if they are on different servers, they can’t damage each other. Let’s fix that.

### Replicate health changes

> In the ThirdPersonShooter project each `TPSCharacter` has a variable tracking health called `CurrentHealth`. The health is deducted on the server when the character is being shot, however as the variable is not setup for replication it does not replicate to the client. Therefore the character's health meter does not update on the client even though the health variable is being modified on the server. 

Reviewer: ```suggestion
In the ThirdPersonShooter project each `TPSCharacter` has a variable tracking health called `CurrentHealth`. The health is deducted on the server when the character is shot, however as the variable is not setup for replication it does not replicate to the client. Therefore the character's health meter does not update on the client even though the health variable is being modified on the server. 
```

3. Back in your two Unreal game clients, run around and shoot.
4. Using the Inspector to track the location of your two players, notice that if you position them in the area of authority then their shots damage each other, but if they are on different servers, they can’t damage each other. Let’s fix that.

### Replicate health changes

> In the ThirdPersonShooter project each `TPSCharacter` has a variable tracking health called `CurrentHealth`. The health is deducted on the server when the character is being shot, however as the variable is not setup for replication it does not replicate to the client. Therefore the characters health meter does not update on the client even though the health variable is being depleted on the server. 

Reviewer: ```suggestion
In the ThirdPersonShooter project each `TPSCharacter` has a variable tracking health called `CurrentHealth`. The health is deducted on the server when the character is being shot, however as the variable is not setup for replication it does not replicate to the client. Therefore the character's health meter does not update on the client even though the health variable is being modified on the server. 
```

To resolve this we need to mark the `CurrentHealth` property for replication as you would with normal [Unreal Actor replication](https://docs.unrealengine.com/en-us/Resources/ContentExamples/Networking/1_1):

Navigate to the declaration of the `CurrentHealth` variable in the `TPSCharacter.h`(line 175) and change the line:
```

> UPROPERTY()

Reviewer: ```suggestion
UPROPERTY(VisibleAnywhere, Category = "Health")
```

```

**NOTE:** We only want to replicate the `CurrentHealth` variable to the client that owns this Actor, thus we specify the `COND_AutonomousOnly` flag.

Finally, we need to implement the `OnRep_CurentHealth` function so that we can update the player health UI when the `CurrentHealth` variable is replicated:

At line 146 in `TPSCharacter.h` add:

```	
UFUNCTION()
void OnRep_CurrentHealth();
```

At line 461 int `TPSCharacter.cpp`, add the following function:

```
void ATPSCharacter::OnRep_CurrentHealth()
{
	if (GetNetMode() != NM_DedicatedServer)
	{
		ATPSPlayerController* PC = Cast<ATPSPlayerController>(GetController());
		if (PC)
		{
			PC->UpdateHealthUI(CurrentHealth, MaxHealth);
		}
		else
		{
			UE_LOG(LogTPS, Warning, TEXT("Couldn't find a player controller for character: %s"), *this->GetName());
		}
	}
}
```

As you probably have noticed, we have just added code exactly as you would in native Unreal Actor replication. The only extra thing we need to do to enable replication is to generate the [schemas]({{urlRoot}}/content/glossary#schema) and the [snapshot]({{urlRoot}}/content/glossary#snapshot) using the Unreal GDK toolbar. 

> Now you can build your game and run. The `CurrentHealth` and the health UI of the actor being shot should go down as intended. 

Reviewer: ```suggestion
Now you can build your game and run. The `CurrentHealth` and the health UI of the Actor being shot should go down as intended. 
```

3. Back in your two Unreal game clients, run around and shoot.
4. Using the Inspector to track the location of your two players, notice that if you position them in the area of authority then their shots damage each other, but if they are on different servers, they can’t damage each other. Let’s fix that.

### Replicate health changes

> In the ThirdPersonShooter project each `TPSCharacter` has a variable tracking health called `CurrentHealth`. The health deducted on the server when the character is being shot, however as the variable is not setup for replication it does not replicate to the client. Therefore the characters health meter does not update on the client even though the health variable is being depleted on the server. 

Reviewer: ```suggestion
In the ThirdPersonShooter project each `TPSCharacter` has a variable tracking health called `CurrentHealth`. The health is deducted on the server when the character is being shot, however as the variable is not setup for replication it does not replicate to the client. Therefore the characters health meter does not update on the client even though the health variable is being depleted on the server. 
```

To resolve this we need to mark the `CurrentHealth` property for replication as you would with normal [Unreal Actor replication](https://docs.unrealengine.com/en-us/Resources/ContentExamples/Networking/1_1):

> Navigate to the declaration of the `CurrentHealth` variable in the `TPSCharacter.h`(line 175) and change the line:

Reviewer 2: Perhaps we shouldn't say "change the line" but rather "add the UProperty specifiers `ReplicatedUsing = OnRep_CurrentHealth`"

Reviewer 2: We need to standardise how we present `go to line`.

Lets go with `TPSCharacter.h - line 175`

So overall this line should be 

Navigate to the declaration of the `CurrentHealth` variable in the `TPSCharacter.h` - line 175 and add the UProperty specifiers `ReplicatedUsing = OnRep_CurrentHealth`:

```
UPROPERTY()
int32 CurrentHealth;
```
To
```
UPROPERTY(ReplicatedUsing = OnRep_CurrentHealth)
int32 CurrentHealth; 
```

We have now marked this property for replication using the `OnRep_CurrentHealth` function which we’ll implement in a moment.

Next we need to update the [GetLifetimeReplicatedProps](https://wiki.unrealengine.com/Replication#Actor_Property_Replication) implementation of the TPSCharacter to specify the [replication conditions](https://docs.unrealengine.com/en-US/Gameplay/Networking/Actors/Properties/Conditions) for the `CurrentHealth` variable:

> Navigate to `TPSCharacter.cpp` line 194 and add the following lines:

Reviewer 2: Perhaps we should specify more detail here rather than just the line number.

Author: Yes that will be for the tidied up tutorial later

```
// Only replicate health to the owning client.

> DOREPLIFETIME_CONDITION(ATPSCharacter, CurrentHealth, COND_AutonomousOnly);

Reviewer: If we're talknig about the owning client, perhaps we should use COND_OwnerOnly.

Author: Added ticket https://improbableio.atlassian.net/browse/UNR-694

```

**NOTE:** We only want to replicate the `CurrentHealth` variable to the client that owns this Actor, thus we specify the `COND_AutonomousOnly` flag.

Finally, we need to implement the `OnRep_CurentHealth` function so that we can update the player health UI when the `CurrentHealth` variable is replicated:

> At line 146 in `TPSCharacter.h` add:

Reviewer 2: Maybe we should change this too `add the public `UFUNCTION` `void OnRep_CurrentHealth();` 
Instead of saying line numbers.

```	
UFUNCTION()
void OnRep_CurrentHealth();
```

> At line 461 int `TPSCharacter.cpp`, add the following function:

Reviewer 2: same as the above. Lets say, in your `TPSCharacter.cpp` add the public function definition: 
Better than line numbers.

```
void ATPSCharacter::OnRep_CurrentHealth()
{
	if (GetNetMode() != NM_DedicatedServer)
	{
		ATPSPlayerController* PC = Cast<ATPSPlayerController>(GetController());
		if (PC)
		{
			PC->UpdateHealthUI(CurrentHealth, MaxHealth);
		}
		else
		{
			UE_LOG(LogTPS, Warning, TEXT("Couldn't find a player controller for character: %s"), *this->GetName());
		}
	}
}
```

> As you probably have noticed, we have just added code exactly as you would in native Unreal Actor replication. The only extra thing we need to do to enable replication is to generate the [schemas]({{urlRoot}}/content/glossary#schema) and the [snapshot]({{urlRoot}}/content/glossary#snapshot) using the Unreal GDK toolbar. 

Reviewer 2: ```suggestion
As you probably have noticed, we have just added code exactly as you would in native Unreal. The only extra thing we need to do to enable replication with the GDK is to generate the [schemas]({{urlRoot}}/content/glossary#schema) and the [snapshot]({{urlRoot}}/content/glossary#snapshot) using the Unreal GDK toolbar. 
```
